core/trino-main/src/main/java/io/trino/operator/PositionSearcher.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.Math.min;
+import static java.lang.Math.toIntExact;
+
+public class PositionSearcher
+{
+    private PositionSearcher()
+    {
+    }
+
+    /**
+     * @param startPosition - inclusive
+     * @param endPosition - exclusive
+     * @param comparator - returns true if positions given as parameters are equal
+     * @return the end of the group position exclusive
+     */
+    public static int findEndPosition(int startPosition, int endPosition, PositionComparator comparator)
+    {
+        checkArgument(startPosition >= 0, "startPosition must be greater or equal than zero: %s", startPosition);
+        checkArgument(startPosition < endPosition, "startPosition (%s) must be less than endPosition (%s)", startPosition, endPosition);
+
+        // exponential search to find the range enclosing the searched position, optimized for small partitions
+        // use long to avoid int overflow
+        long left = startPosition;
+        long right = startPosition;
+        long distance = 1;
+        while (right < endPosition && comparator.test(toIntExact(left), toIntExact(right))) {
+            left = right;
+            right += distance;
+            distance *= 2;
+        }
+        // binary search to find the searched position within the range
+        int intLeft = toIntExact(left);
+        int intRight = toIntExact(min(right, endPosition));
```

Comment:
Since `endPosition` is exclusive, `endPosition - 1` ? It works either way because of how `middle` is computed below. But the latter is clearer and more "correct", as technically, the `intRight` could be used as the middle position.

Reply:
We return `intRight` from this method. It is supposed to be the first position that differs. If all positions in the partition are equal, we return the first position after the partition. if we moved `intRight` to the left, we would not be able to do it.

Reply:
Then return `intRight + 1` ?